### PR TITLE
Add middleware for new relic instrumentation

### PR DESCRIPTION
**Pull request by sanjayprabhu**
Integrates with https://github.com/remind101/nra for new relic transaction tracing.

TODO
- Test with newrelic
- Figure out build. This is going to break the build on linux because this depends on the new relic agent sdk. I'm thinking about having the nra package build with a no-op implementation by default, but you could optionally link to the new relic agent if you wanted to. Not sure if that will work though. 
- Maybe add to `example/main.go`

cc @ejholmes @bmarini

### CODE REVIEW

File: httpx/middleware/newrelic_tracer.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	tx := h.createTx(transactionName, r.URL.String(), h.tracer)
+	ctx = nra.WithTx(ctx, tx)
+
+	tx.Start()
+	defer tx.End()
+
+	return h.handler.ServeHTTPContext(ctx, w, r)
+}
+
+func (h *NewRelicTracer) templatePath(r *http.Request) string {
```
Comment (ejholmes) [inline httpx/middleware/newrelic_tracer.go]: Minor style thing, but in general, I think it's good to pull stuff like this into a func if it doesn't actually use anything from the struct and it's not there to satisfy an interface. Pulling it into a func makes it easier to test without having to setup a NewRelicTracer (if you wanted to unit test it).

File: httpx/middleware/newrelic_tracer.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	createTx func(string, string, nra.TxTracer) nra.Tx
+}
+
+func NewRelicTracing(h httpx.Handler, router *httpx.Router, tracer nra.TxTracer) *NewRelicTracer {
+	return &NewRelicTracer{h, tracer, router, createTx}
+}
+
+func (h *NewRelicTracer) ServeHTTPContext(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	templatePath := h.templatePath(r)
+	transactionName := fmt.Sprintf("%s (%s)", templatePath, r.Method)
```
Comment (ejholmes) [inline httpx/middleware/newrelic_tracer.go]: What about `METHOD /path` or `METHOD: /path`? I think that's what we use in API. Feels a little more consistent with the http spec.

File: httpx/middleware/newrelic_tracer_test.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+			routes: func(r *httpx.Router) {
+				r.Handle("/path", httpx.HandlerFunc(func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+					return nil
+				})).Methods("GET")
+			},
+			req: newRequest("GET", "/path"),
+			expectedTransactionName: "/path (GET)",
+			expectedUrl:             "/path",
+		},
+		// path with variables
```
Comment (ejholmes) [inline httpx/middleware/newrelic_tracer_test.go]: Just for sanity, might wanna add a test with a regex param like `/articles/{category}/{id:[0-9]+}`.

File: httpx/middleware/newrelic_tracer.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"fmt"
+	"github.com/remind101/nra"
+	"github.com/remind101/pkg/httpx"
+	"golang.org/x/net/context"
+	"net/http"
+)
+
+type NewRelicTracer struct {
```
Comment (ejholmes) [inline httpx/middleware/newrelic_tracer.go]: In the docs for this, might wanna mention that this needs to be added downstream from the httpx.Router; so wrapping an httpx.Router won't work, you need to add it when adding a handler: `r.Handle(NewRelicTracing(h, ...))`.

**Comment by ejholmes** [comment]
I like it! Travis didn't run the tests for some reason, but otherwise lgtm :+1: 

I wonder if this belongs in another package so we don't add the nra dependency (could add a middleware package in nra with this `github.com/remind101/nra/middleware`)

**Comment by bmarini** [comment]
I've opened this issue for better support from gorilla/mux, but this PR will suffice for now I think: https://github.com/gorilla/mux/issues/97

**Comment by ejholmes** [comment]
:+1:

**Pull request merged**